piptools/scripts/compile.py

```diff
@@ -270,29 +278,46 @@ def cli(
                 ).format(DEFAULT_REQUIREMENTS_FILE)
             )
 
+    src_files = tuple(src if src == "-" else os.path.abspath(src) for src in src_files)
+
+    ###
+    # Resolve output file
+    ###
+
     if not output_file:
+        # An output file must be provided if there are multiple source files
+        if len(src_files) > 1:
+            raise click.BadParameter(
+                "--output-file is required if two or more input files are given."
+            )
         # An output file must be provided for stdin
-        if src_files == ("-",):
+        elif src_files == ("-",):
             raise click.BadParameter("--output-file is required if input is from stdin")
-        # Use default requirements output file if there is a setup.py the source file
+        # Use default requirements output file if there is only a setup.py source file
         elif os.path.basename(src_files[0]) == "setup.py":
             file_name = os.path.join(
                 os.path.dirname(src_files[0]), DEFAULT_REQUIREMENTS_OUTPUT_FILE
             )
-        # An output file must be provided if there are multiple source files
-        elif len(src_files) > 1:
-            raise click.BadParameter(
-                "--output-file is required if two or more input files are given."
-            )
         # Otherwise derive the output file from the source file
         else:
-            base_name = src_files[0].rsplit(".", 1)[0]
-            file_name = base_name + ".txt"
+            file_name = os.path.splitext(src_files[0])[0] + ".txt"
+            if file_name == src_files[0]:
+                file_name += ".txt"
+    else:
+        file_name = output_file
+
+    if file_name != "-":
+        file_name = os.path.abspath(file_name)
+        # Ensure source paths are relative to the output file:
+        os.chdir(os.path.dirname(file_name))
```

Comment:
Could something like `with chdir(...)` (or a decorator) be used to make sure that we return to the original dir if something went wrong?

Reply:
Yes, it's a good idea.

But the primary problem right now is figuring out exactly when we need to be in each of the `output_dir` and `invocation_dir`, if that's even going to work. If it won't we might need to do more path rewriting... I need to try some more things.

While the current state of this PR is passing existing tests, here's an example of what's wrong:

```console
$ git clone git@github.com:jazzband/pip-tools
$ cd pip-tools
$ mkdir subdir
$ echo 'file:.' >requirements.in
$ pip-compile -o subdir/reqs.txt requirements.in
```

<details><summary>Failure to find the package:</summary>

```
      ERROR: Command errored out with exit status 1:
       command: /home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-kzbbcwdh/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-kzbbcwdh/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-iq5mc42e
           cwd: /tmp/pip-req-build-kzbbcwdh/
      Complete output (5 lines):
      Traceback (most recent call last):
        File "<string>", line 1, in <module>
        File "/home/andy/.pyenv/versions/3.8.7/lib/python3.8/tokenize.py", line 392, in open
          buffer = _builtin_open(filename, 'rb')
      FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-req-build-kzbbcwdh/setup.py'
      ----------------------------------------
Traceback (most recent call last):
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/bin/pip-compile", line 33, in <module>
    sys.exit(load_entry_point('pip-tools', 'console_scripts', 'pip-compile')())
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/andy/Code/pip-tools/piptools/scripts/compile.py", line 475, in cli
    results = resolver.resolve(max_rounds=max_rounds)
  File "/home/andy/Code/pip-tools/piptools/resolver.py", line 184, in resolve
    has_changed, best_matches = self._resolve_one_round()
  File "/home/andy/Code/pip-tools/piptools/resolver.py", line 264, in _resolve_one_round
    constraints = sorted(self.constraints, key=key_from_ireq)
  File "/home/andy/Code/pip-tools/piptools/resolver.py", line 136, in constraints
    return set(
  File "/home/andy/Code/pip-tools/piptools/resolver.py", line 240, in _group_constraints
    self.repository.get_dependencies(ireq)
  File "/home/andy/Code/pip-tools/piptools/repositories/pypi.py", line 243, in get_dependencies
    self._dependencies_cache[ireq] = self.resolve_reqs(
  File "/home/andy/Code/pip-tools/piptools/repositories/pypi.py", line 201, in resolve_reqs
    results = resolver._resolve_one(reqset, ireq)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 388, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 340, in _get_abstract_dist_for
    abstract_dist = self.preparer.prepare_linked_requirement(req)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/operations/prepare.py", line 482, in prepare_linked_requirement
    abstract_dist = _get_prepared_distribution(
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/operations/prepare.py", line 91, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/distributions/sdist.py", line 40, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/req/req_install.py", line 555, in prepare_metadata
    self.metadata_directory = self._generate_metadata()
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/req/req_install.py", line 530, in _generate_metadata
    return generate_metadata_legacy(
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/operations/build/metadata_legacy.py", line 70, in generate_metadata
    call_subprocess(
  File "/home/andy/.local/share/venvs/1b2b5c9cc79c4656dd687f3b665b224c/venv-python-3.8.7/lib/python3.8/site-packages/pip/_internal/utils/subprocess.py", line 242, in call_subprocess
    raise InstallationError(exc_msg)
pip._internal.exceptions.InstallationError: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.

```
</details>


I will add a test like this and then work to get it passing.

Sorry for the early review alerts.

Reply:
I've reduced the actual occurrence of `chdir`, and when it is used, I'm now wrapping it in a simple `try:`+`finally:` block, rather than creating a more specific decorator.

`plumbum` provides very useful context managers for this kind of stuff, but I don't think this project's current path tools can say the same.

Reply:
Added a real context manager, now that there's a second use for that pattern, thanks!